internal/query/relations.go

```diff
@@ -267,19 +269,187 @@ func (e *engine) CreateRelationships(ctx context.Context, rels []types.Relations
-func (e *engine) CreateRole(ctx context.Context, res types.Resource, actions []string) (types.Role, error) {
-	role := newRole(actions)
+func (e *engine) CreateRole(ctx context.Context, actor, res types.Resource, roleName string, actions []string) (types.Role, error) {
+	ctx, span := e.tracer.Start(ctx, "engine.CreateRole")
+
+	defer span.End()
+
+	roleName = strings.TrimSpace(roleName)
+
+	role := newRole(roleName, actions)
 	roleRels := e.roleRelationships(role, res)
 
+	dbCtx, err := e.store.BeginContext(ctx)
+	if err != nil {
+		return types.Role{}, nil
+	}
+
+	dbRole, err := e.store.CreateRole(dbCtx, actor.ID, role.ID, roleName, res.ID)
+	if err != nil {
+		return types.Role{}, err
+	}
+
 	request := &pb.WriteRelationshipsRequest{Updates: roleRels}
 
 	if _, err := e.client.WriteRelationships(ctx, request); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		return types.Role{}, err
+	}
+
+	if err = e.store.CommitContext(dbCtx); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
 		return types.Role{}, err
 	}
 
+	role.CreatorID = dbRole.CreatorID
+	role.ResourceID = dbRole.ResourceID
+	role.CreatedAt = dbRole.CreatedAt
+	role.UpdatedAt = dbRole.UpdatedAt
+
 	return role, nil
 }
 
+// actionsDiff determines which actions needs to be added and removed.
+// If no new actions are provided it is assumed no changes are requested.
+func actionsDiff(oldActions, newActions []string) ([]string, []string) {
+	if len(newActions) == 0 {
+		return nil, nil
+	}
+
+	old := make(map[string]bool, len(oldActions))
+	new := make(map[string]bool, len(newActions))
+
+	var add, rem []string
+
+	for _, action := range oldActions {
+		old[action] = true
+	}
+
+	for _, action := range newActions {
+		new[action] = true
+
+		// If the new action is not in the old actions, then we need to add the action.
+		if !old[action] {
+			add = append(add, action)
+		}
+	}
+
+	for _, action := range oldActions {
+		// If the old action is not in the new actions, then we need to remove it.
+		if !new[action] {
+			rem = append(rem, action)
+		}
+	}
+
+	return add, rem
+}
+
+// UpdateRole allows for updating an existing role with a new name and new actions.
+// If new name is empty, no change is made.
+// If new actions is an empty slice, no change is made.
+func (e *engine) UpdateRole(ctx context.Context, actor, roleResource types.Resource, newName string, newActions []string) (types.Role, error) {
+	ctx, span := e.tracer.Start(ctx, "engine.UpdateRole")
+
+	defer span.End()
+
+	role, err := e.GetRole(ctx, roleResource)
+	if err != nil {
+		return types.Role{}, err
+	}
+
+	newName = strings.TrimSpace(newName)
+
+	// If the name is the same, then clear the changed name.
+	if role.Name == newName {
+		newName = ""
+	}
```

Comment:
good point. i'll update it to instead, if no name is provided (for example, when just updating actions) it will populate the name from the existing record so we can keep the query simple.